File: MedianOfTwo/MedianOfTwo.c
#include <stdio.h>
#include <limits.h>

double findMedianSortedArrays(int* nums1, int m, int* nums2, int n) {
    // Ensure nums1 is the shorter array for efficiency
    if (m > n) {
        return findMedianSortedArrays(nums2, n, nums1, m);
    }

    int low = 0;
    int high = m;

    while (low <= high) {
        int partitionX = (low + high) / 2;
        int partitionY = ((m + n + 1) / 2) - partitionX;

        int maxLeftX = (partitionX == 0) ? INT_MIN : nums1[partitionX - 1];
        int minRightX = (partitionX == m) ? INT_MAX : nums1[partitionX];

        int maxLeftY = (partitionY == 0) ? INT_MIN : nums2[partitionY - 1];
        int minRightY = (partitionY == n) ? INT_MAX : nums2[partitionY];

        if (maxLeftX <= minRightY && maxLeftY <= minRightX) {
            // Found the correct partition
            if ((m + n) % 2 == 0) {
                // Even number of elements
                return (double)(fmax(maxLeftX, maxLeftY) + fmin(minRightX, minRightY)) / 2;
            } else {
                // Odd number of elements
                return (double)fmax(maxLeftX, maxLeftY);
            }
        } else if (maxLeftX > minRightY) {
            // Need to move partitionX to the left
            high = partitionX - 1;
        } else {
            // Need to move partitionX to the right
            low = partitionX + 1;
        }
    }

    // This point should never be reached if the arrays are sorted
    return -1.0; 
}

File: MedianOfTwo/MedianOfTwo.rs
impl Solution {
    pub fn find_median_sorted_arrays(nums1: Vec<i32>, nums2: Vec<i32>) -> f64 {
        let m = nums1.len();
        let n = nums2.len();

        // Ensure nums1 is the shorter array for efficiency
        if m > n {
            return Solution::find_median_sorted_arrays(nums2, nums1);
        }

        let mut low = 0;
        let mut high = m;

        while low <= high {
            let partition_x = (low + high) / 2;
            let partition_y = ((m + n + 1) / 2) - partition_x;

            let max_left_x = if partition_x == 0 {
                i32::MIN
            } else {
                nums1[partition_x - 1]
            };

            let min_right_x = if partition_x == m {
                i32::MAX
            } else {
                nums1[partition_x]
            };

            let max_left_y = if partition_y == 0 {
                i32::MIN
            } else {
                nums2[partition_y - 1]
            };

            let min_right_y = if partition_y == n {
                i32::MAX
            } else {
                nums2[partition_y]
            };

            if max_left_x <= min_right_y && max_left_y <= min_right_x {
                // Found the correct partition
                if (m + n) % 2 == 0 {
                    // Even number of elements
                    return (f64::max(max_left_x as f64, max_left_y as f64)
                        + f64::min(min_right_x as f64, min_right_y as f64))
                        / 2.0;
                } else {
                    // Odd number of elements
                    return f64::max(max_left_x as f64, max_left_y as f64);
                }
            } else if max_left_x > min_right_y {
                // Need to move partition_x to the left
                high = partition_x - 1;
            } else {
                // Need to move partition_x to the right
                low = partition_x + 1;
            }
        }

        unreachable!("The input arrays must be sorted.");
    }
}